crates/sui-indexer/src/handlers/checkpoint_handler_v2.rs

```diff
@@ -0,0 +1,647 @@
+        .unwrap_or(CHECKPOINT_QUEUE_SIZE.to_string())
+        .parse::<usize>()
+        .unwrap();
+    let global_metrics = get_metrics().unwrap();
+    let (indexed_checkpoint_sender, indexed_checkpoint_receiver) =
+        mysten_metrics::metered_channel::channel(
+            checkpoint_queue_size,
+            &global_metrics
+                .channels
+                .with_label_values(&["checkpoint_indexing"]),
+        );
+
+    let state_clone = state.clone();
+    let metrics_clone = metrics.clone();
+    let config_clone = config.clone();
+    let (tx, rx) = watch::channel(None);
+    spawn_monitored_task!(start_tx_checkpoint_commit_task(
+        state_clone,
+        metrics_clone,
+        config_clone,
+        indexed_checkpoint_receiver,
+        tx,
+    ));
+
+    let checkpoint_processor = CheckpointHandler {
+        state,
+        metrics,
+        indexed_checkpoint_sender,
+        checkpoint_starting_tx_seq_numbers: HashMap::new(),
+        package_cache: IndexingPackageCache::start(rx),
+    };
+
+    Ok(checkpoint_processor)
+}
+
+pub struct CheckpointHandler<S> {
+    state: S,
+    metrics: IndexerMetrics,
+    indexed_checkpoint_sender: mysten_metrics::metered_channel::Sender<CheckpointDataToCommit>,
+    // Map from checkpoint sequence number and its starting transaction sequence number
+    // This thing is small enough to be kept in memory
+    checkpoint_starting_tx_seq_numbers: HashMap<CheckpointSequenceNumber, u64>,
+    package_cache: Arc<Mutex<IndexingPackageCache>>,
+}
+
+#[async_trait]
+impl<S> Handler for CheckpointHandler<S>
+where
+    S: IndexerStoreV2 + Clone + Sync + Send + 'static,
+{
+    fn name(&self) -> &str {
+        "checkpoint-handler"
+    }
+
+    async fn process_checkpoint(&mut self, checkpoint_data: &CheckpointData) -> anyhow::Result<()> {
+        let checkpoint_seq = checkpoint_data.checkpoint_summary.sequence_number();
+        info!(checkpoint_seq, "Checkpoint received by CheckpointHandler");
+
+        // update next checkpoint starting tx seq number
+        // Note transaction is 0-indexed, so do epoch and checkpoint
+        self.checkpoint_starting_tx_seq_numbers.insert(
+            *checkpoint_seq + 1,
+            checkpoint_data
+                .checkpoint_summary
+                .network_total_transactions,
+        );
+        let current_checkpoint_starting_tx_seq = if checkpoint_seq == &0 {
+            0
+        } else if self
+            .checkpoint_starting_tx_seq_numbers
+            .contains_key(checkpoint_seq)
+        {
+            self.checkpoint_starting_tx_seq_numbers[checkpoint_seq]
+        } else {
+            self.state.get_checkpoint_ending_tx_sequence_number(checkpoint_seq - 1).await?
+            .unwrap_or_else(|| {
+                panic!("While processing checkpoint {}, we failed to find the starting tx seq both in mem and DB.", checkpoint_seq)
+            }) + 1
+        };
```

Comment:
I don't think we need to know the previous checkpoints ending tx sequence number in order to calculate the transaction sequence number for each transaction in the current checkpoint we're processing. See https://mystenlabs.github.io/sui/sui_types/messages_checkpoint/enum.CheckpointContents.html#method.enumerate_transactions for more on this.

Reply:
oh this is very cool, I didn't know about it. Let me see how I can use it here

Reply:
fixed. This is much better than what I had initially. Thanks!